text/0000-try-catch.md

```diff
@@ -0,0 +1,98 @@
+- Feature Name: try_catch
+- Start Date: 2020-05-21
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new `Context` method for catching any JavaScript exceptions that may be thrown during the execution of a computation. The result of the API uses Rust's standard `Result` enum to distinguish between normal return values and exceptional values:
+```rust
+match cx.try_catch(|cx| { cx.throw(42) }) {
+    Ok(v) => { /* ... computation produced a normal result ... */ }
+    Err(v) => { /* ... computation threw an exception ... */ }
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Today, Neon APIs that throw JavaScript exceptions product a Rust `Result` with the `Throw` sentinel value, which simply indicates that the JS VM has entered into a throwing state. But they do not offer a way to model the JavaScript `catch` semantics, which restores the VM to non-throwing state and extracts the exceptional value that was being thrown.
+
+This RFC proposes an idiomatic way to mimic JavaScript's `try`/`catch` syntax with a Neon API, which "tries" a computation encapsulated in a Rust callback and "catches" any exception that gets thrown and reflects that in an idiomatic Rust `Result`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Throwing vs non-throwing state
+
+The JavaScript VM can be in a _throwing state_ or _non-throwing state_. Any API that can trigger the throwing of a JavaScript exception value sets the VM into throwing state. **When the VM is in a throwing state, most Neon APIs are not allowed to be called and will panic.**
+
+Every Neon API is documented as either _safe for throwing state_, meaning that the API can safely be called without panicking even when the VM is in a throwing state, or _unsafe for throwing state_, meaning that it will panic if called when the VM is in a throwing state.
+
+## Catching exceptions
+
+The `try_catch` method runs a callback and intercepts any thrown exception value and ensures the JavaScript VM is restored to non-throwing state.
+
+Example:
+
+```rust
+let v = match cx.try_catch(|cx| { cx.throw(v) }) {
+    Ok(v) => v,
+    Err(v) => v
+};
+```
+
+## Inspecting the exception state
+
+At any time, the VM can be inspected to detect whether it is a throwing or non-throwing state:
+
+```rust
+cx.is_throwing()
+```
+
+The `is_throwing()` method is safe for throwing state.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `neon::context::Context` trait gets two new methods:
+
+```rust
+pub trait Context<'a> {
+    fn is_throwing(&self) -> bool;
+    fn try_catch<T, U, F>(&self, f: F) -> Result<Handle<'a, T>, Handle<'a, U>>
+    where
+        T: Value,
+        U: Value,
+        F: for<'b> FnOnce(CatchContext<'b>) -> JsResult<'b, T>;
```

Comment:
A try/catch block doesn't change `this` bindings or anything else contextual in Js. Instead of a `CatchContext` should `F` be generic over the context that called `try_catch`?

```rust
pub trait Context<'a> {
    fn try_catch<T, U, F>(&self, f: F) -> Result<Handle<'a, T>, Handle<'a, U>>
    where
        T: Value,
        U: Value,
        F: for<'b> FnOnce(Self) -> JsResult<'b, T>;
}
```

Reply:
If we can just eliminate the context and let the closure refer to the outer context, that seems strictly more convenient. It'll probably require implementation to be sure, but for now I'll remove the `CatchContext` type and the context argument.

Reply:
We weren't able to eliminate the `cx` argument entirely because of Rust's type rules around `&mut` references. But I did take your suggestion and use `Self` as the type of the inner context and just pass `self` in directly.